lib/Runtime/Debug/TTSnapValues.cpp

```diff
@@ -95,53 +82,76 @@ namespace TTD
             }
             else
             {
-                into.Contents = HeapNewArray(wchar, into.Length + 1);
-                js_memcpy_s(into.Contents, into.Length * sizeof(wchar), string, length * sizeof(wchar));
-                into.Contents[into.Length] = '\0';
+                into.Contents = TT_HEAP_ALLOC_ARRAY(char16, into.Length + 1);
+                js_memcpy_s(into.Contents, into.Length * sizeof(char16), string, length * sizeof(char16));
+                into.Contents[into.Length] = _u('\0');
             }
         }
 
         void DeleteStringFromHeapAllocator(TTString& string)
         {
             if(string.Contents != nullptr)
             {
-                HeapDeleteArray(string.Length + 1, string.Contents);
+                TT_HEAP_FREE_ARRAY(char16, string.Contents, string.Length + 1);
             }
         }
 
-        void WriteCodeToFile(IOStreamFunctions& streamFunctions, LPCWSTR srcDir, LPCWSTR docId, LPCWSTR sourceUri, const wchar* sourceBuffer, uint32 length)
+        void WriteCodeToFile(ThreadContext* threadContext, bool fromEvent, DWORD_PTR docId, bool isUtf8Source, byte* sourceBuffer, uint32 length)
         {
-            HANDLE srcStream = streamFunctions.pfGetSrcCodeStream(srcDir, docId, sourceUri, false, true);
+            char asciiResourceName[64];
+            sprintf_s(asciiResourceName, 64, "src%s_%I64u.js", (fromEvent ? "_ld" : ""), static_cast<uint64>(docId));
 
-            byte byteOrderArray[2] = { 0xFF, 0xFE };
-            DWORD byteOrderCount = 0;
-            bool okBOC = streamFunctions.pfWriteBytesToStream(srcStream, byteOrderArray, 2, &byteOrderCount);
-            AssertMsg(okBOC && byteOrderCount == 2, "Write Failed!!!");
+            JsTTDStreamHandle srcStream = threadContext->TTDStreamFunctions.pfGetResourceStream(threadContext->TTDUri.UriByteLength, threadContext->TTDUri.UriBytes, asciiResourceName, false, true);
+
+            if(isUtf8Source)
+            {
+                byte byteOrderArray[3] = { 0xEF, 0xBB, 0xBF };
+                size_t byteOrderCount = 0;
+                bool okBOC = threadContext->TTDStreamFunctions.pfWriteBytesToStream(srcStream, byteOrderArray, 3, &byteOrderCount);
+                AssertMsg(okBOC && byteOrderCount == 3, "Write Failed!!!");
```

Comment:
Only byteOrderArray is different between if and else, can we refactor and keep the call to pfWriteBytesToStream and assert as common? can use _countof(byteOrderArray) for size. Same thing for read?